step-04/index.js

```diff
@@ -29,15 +29,15 @@ io.sockets.on('connection', function(socket) {
   socket.on('create or join', function(room) {
     log('Received request to create or join room ' + room);
 
-    var numClients = io.sockets.sockets.length;
+    var clientsInRoom = io.sockets.adapter.rooms[room];
```

Comment:
To illustrate what I mean:
```bash
$ node index.js

# First user creates a room, called `testroom`.
{ '-AzFZXmaqM8aEl5KAAAC': { '-AzFZXmaqM8aEl5KAAAC': true },
  testroom: { '-AzFZXmaqM8aEl5KAAAC': true } }
# Second user joins `testroom`.
{ '-AzFZXmaqM8aEl5KAAAC': { '-AzFZXmaqM8aEl5KAAAC': true },
  testroom: { '-AzFZXmaqM8aEl5KAAAC': true, lomYavt1vwPDv9mRAAAD: true },
  lomYavt1vwPDv9mRAAAD: { lomYavt1vwPDv9mRAAAD: true } }
# Third user attempts to join `testroom` but is rejected due to limit.
{ '-AzFZXmaqM8aEl5KAAAC': { '-AzFZXmaqM8aEl5KAAAC': true },
  testroom: { '-AzFZXmaqM8aEl5KAAAC': true, lomYavt1vwPDv9mRAAAD: true },
  lomYavt1vwPDv9mRAAAD: { lomYavt1vwPDv9mRAAAD: true },
  'N3sh8wcByW7EanT-AAAE': { 'N3sh8wcByW7EanT-AAAE': true } }
```

What I have done here is the following.

```js
socket.on('create or join', function(room) {
  log('Received request to create or join room ' + room);

  var clientsInRoom = io.sockets.adapter.rooms[room];
  var numClients = clientsInRoom ? Object.keys(clientsInRoom).length : 0;
  log('Room ' + room + ' now has ' + numClients + ' client(s)');

  if (numClients === 0) {
    socket.join(room);
    log('Client ID ' + socket.id + ' created room ' + room);
    socket.emit('created', room, socket.id);
  } else if (numClients === 1) {
    log('Client ID ' + socket.id + ' joined room ' + room);
    io.sockets.in(room).emit('join', room);
    socket.join(room);
    socket.emit('joined', room, socket.id);
    io.sockets.in(room).emit('ready');
  } else { // max two clients
    socket.emit('full', room);
  }
  // This generates the output shown in previous example.
  console.log(io.sockets.adapter.rooms) 
});
```

As you can see, the current solution works for me and each user is assigned a unique room corresponding to their unique ID.